perma_web/frontend/components/FolderSelect.vue

```diff
@@ -0,0 +1,159 @@
+<script setup>
+import { computed, ref } from 'vue'
+import { globalStore } from '../stores/globalStore'
+import { onClickOutside } from '@vueuse/core'
+
+const selectContainerRef = ref(null)
+const selectButtonRef = ref(null)
+const selectListRef = ref(null)
+const isSelectExpanded = ref(false)
+const selectLabel = computed(() => globalStore.selectedFolder.path.length ? globalStore.selectedFolder.path.join(" > ") : 'Please select a folder')
+
+const folders = computed(() => globalStore.organizationFolders.concat(globalStore.sponsoredFolders))
+
+// Required for admin users locally, where admins have an "empty root folder" instead of personal links
+const personalFolderNames = ['Empty root folder', 'Personal Links']
+
+const personalFolderId = current_user.top_level_folders[0].id
+
+const getFolderHeader = (folder) => {
+    if (folder.registrar) {
+        return folder.registrar
+    }
+
+    else if (folder.sponsored_by) {
+        return "Sponsored Links"
+    }
+
+    return "Personal Links"
+}
+
+const showLinksRemaining = computed(() => personalFolderNames.includes(globalStore.selectedFolder.path[0]) || !!globalStore.selectedFolder.isReadOnly)
+const linksRemaining = computed(() => {
+    if (globalStore.selectedFolder.isReadOnly) {
+        return 0
+    }
+
+    else if (globalStore.linksRemaining === Infinity) {
+        return "unlimited"
+    }
+
+    return globalStore.linksRemaining
+})
+
+// Select Event Handlers
+onClickOutside(selectContainerRef, () => {
+    if (!isSelectExpanded) { return }
+    isSelectExpanded.value = false
+})
+
+const handleSelectToggle = () => {
+    isSelectExpanded.value = !isSelectExpanded.value
+}
+
+const handleFocus = (index) => {
+    const itemToFocus = selectListRef.value.querySelector(`[data-index="${index}"]`);
+    itemToFocus.focus();
+};
+
+const handleArrowDown = (e) => {
+    const currentIndex = parseInt(e.srcElement.dataset?.index)
+
+    if (!currentIndex) {
+        handleFocus(0)
+    }
+
+    if (currentIndex < folders.value.length) {
+        handleFocus(currentIndex + 1)
+    }
+}
+
+const handleArrowUp = (e) => {
+    const currentIndex = parseInt(e.srcElement.dataset.index)
+    if (currentIndex > 0) {
+        handleFocus(currentIndex - 1)
+    }
+}
+
+const handleClose = () => {
+    isSelectExpanded.value = false
+    selectButtonRef.value.focus()
+}
+
+const handleSelection = (e) => {
+    const isSpan = e.target.matches('span')
+    const target = isSpan ? e.target.parentElement : e.target
+
+    const { orgid, folderid } = target.dataset
```

Comment:
In this stanza, I'm having a trouble reading through, keeping track of the difference between `orgid` and `orgId`, and similarly with `folderid` and `folderId`. I'm not sure what to suggest, but I'm wondering if there might be a way to rewrite this section such that such similar variable names aren't necessary. Do you think using `target.dataset['orgid']` and `target.dataset['folderid']` rather than nicknaming might be clearer? Or possibly doing the parsing inline, right inside the event definition? 🤷‍♀️ Do you have any thoughts here?

Reply:
For sure, I definitely wish data-attributes didn't require all lowercase names and allow us to avoid this. I think there are two possible approaches that might make this easier to read: 

1. We could rename the variables inline at the same time they are destructured from `target.dataset` (and entirely avoid the need for the additionally declared variables)  

```js
const { orgid: orgId, folderid: folderId } = target.dataset

    if (!folderId) {
        return handleClose()
    }

    // Call a custom event that triggers triggerOnWindow function
    const updateSelections = new CustomEvent("dropdown.selectionChange", { detail: { data: { folderId: JSON.parse(folderId), orgId: orgId ? parseInt(orgId) : null } } });
    window.dispatchEvent(updateSelections);
```

Renaming variables at the same time they are destructured like this is pretty common, and pretty concise. 

2. But another totally reasonable option is to not use destructuring at all, like you mentioned:  
```js
    if (!target.dataset.folderid) {
        return handleClose()
    }
    
    const updateSelections = new CustomEvent("dropdown.selectionChange", { detail: { data: { folderId: JSON.parse(target.dataset.folderid), orgId: target.dataset.orgid ? parseInt(target.dataset.orgid) : null } } });
    window.dispatchEvent(updateSelections);
```

Which do you find easier to read?

Reply:
Great, this version seems much easier to follow to me! Thanks!